build/curriculum.ts

```diff
@@ -0,0 +1,422 @@
+    forIndex?: boolean;
+  }
+): Promise<ReadCurriculum> {
+  const raw = await fs.readFile(file, "utf-8");
+  const { attributes, body: rawBody } = frontmatter<CurriculumFrontmatter>(raw);
+  const filename = file.replace(CURRICULUM_ROOT, "").replace(/^\/?/, "");
+  let title = rawBody.match(/^[\w\n]*#+(.*\n)/)[1]?.trim();
+  const body = rawBody.replace(/^[\w\n]*#+(.*\n)/, "");
+
+  const slug = fileToSlug(file);
+  const url = `/${DEFAULT_LOCALE}/${slug}/`;
+
+  let sidebar: CurriculumIndexEntry[];
+  let parents: DocParent[];
+
+  let modules: CurriculumIndexEntry[];
+  let prevNext: PrevNext;
+  if (!options?.forIndex) {
+    if (attributes.template === Template.landing) {
+      modules = (await buildCurriculumIndex())?.filter(
+        (x) => x.children?.length
+      );
+    } else if (attributes.template === Template.overview) {
+      modules = (await buildCurriculumIndex())?.find(
+        (x) => x.slug === slug
+      )?.children;
+    }
+    if (attributes.template === Template.module) {
+      prevNext = await buildPrevNextModule(slug);
+    } else if (attributes.template === Template.overview) {
+      prevNext = await buildPrevNextOverview(slug);
+    }
+
+    sidebar = await buildCurriculumSidebar();
+    parents = await buildParents(url);
+  } else {
+    title = title
+      .replace(/^\d+\s+/, "") // Strip number prefix.
+      .replace(/ modules$/, "") // Strip "modules" suffix.
+      .replace(/Extension \d+:/, ""); // Strip "Extension" prefix.
+  }
+
+  return {
+    meta: {
+      filename,
+      slug,
+      url,
+      title,
+      sidebar,
+      modules,
+      parents,
+      prevNext,
+      ...attributes,
+    },
+    body,
+  };
+}
+
+export async function findCurriculumPageBySlug(
+  slug: string
+): Promise<CurriculumData | null> {
+  let file = await slugToFile(slug);
+  if (!file) {
+    file = await slugToFile(`${slug}${slug ? "/" : ""}README`);
+  }
+  let module;
+  try {
+    module = await readCurriculumPage(file, { forIndex: false });
+  } catch (e) {
+    console.error(`No file found for ${slug}: ${e}`);
+    return;
+  }
+  const { body, meta } = module;
+
+  const d: CurriculumBuildData = {
+    url: meta.url,
+    rawBody: body,
+    metadata: { locale: DEFAULT_LOCALE, ...meta },
+    isMarkdown: true,
+    fileInfo: {
+      path: file,
+    },
+  };
+
+  const doc = await buildCurriculumPage(d);
+  return { doc };
+}
+
+export async function buildCurriculumPage(
+  document: CurriculumBuildData
+): Promise<Doc> {
+  const { metadata } = document;
+
+  const doc = { locale: DEFAULT_LOCALE } as Partial<CurriculumDoc>;
+  let $ = null;
+
+  const renderUrl = document.url.replace(/\/$/, "");
+  [$] = await kumascript.render(renderUrl, {}, document as any);
+
+  $("[data-token]").removeAttr("data-token");
+  $("[data-flaw-src]").removeAttr("data-flaw-src");
+
+  doc.title = (metadata.title || "").replace(/^\d+\s+/, "");
+  doc.mdn_url = document.url;
+  doc.locale = metadata.locale as string;
```

Comment:
Is this assertion necessary?

```suggestion
  doc.locale = metadata.locale;
```